File: PathTerminalSDK/Sources/PathEmulatorAdapter/EmulatorWireJsonMapping.swift
/// Maps Nordic UART newline-delimited JSON `result` objects to ``TransactionResult``.
/// Shared so unit tests can validate fixtures without BLE hardware.
import Foundation
import PathCoreModels

enum EmulatorWireJsonMapping {
    static func mapTransactionResult(
        _ raw: [String: Any],
        defaultState: TransactionState,
        requestId: String,
        amountFallback: Int,
        currencyFallback: String
    ) -> TransactionResult {
        let topStatus = (raw["status"] as? String)?.lowercased() ?? ""
        let txnStatus = (raw["txn_status"] as? String) ?? (raw["transaction_status"] as? String)
        let paymentStatus: String? = {
            if let t = txnStatus, !t.isEmpty { return t.lowercased() }
            if topStatus == "success" || topStatus == "error" { return nil }
            return topStatus.isEmpty ? nil : topStatus
        }()

        let state: TransactionState
        let error: PathError?

        if topStatus == "error" {
            let msg = raw["error"] as? String ?? "Terminal error"
            state = .declined
            error = PathError(code: .decline, message: msg, recoverable: false)
        } else if let ps = paymentStatus, !ps.isEmpty {
            switch ps {
            case "approved", "success":
                state = defaultState
                error = nil
            case "refunded":
                state = .refunded
                error = nil
            case "cancelled", "canceled":
                state = .cancelled
                error = nil
            case "timed_out", "timedout":
                state = .timedOut
                error = PathError(code: .timeout, message: raw["error"] as? String ?? "Transaction timed out", recoverable: true)
            case "declined":
                state = .declined
                error = PathError(code: .decline, message: raw["error"] as? String ?? "Declined", recoverable: false)
            case "processing", "pending", "pending_device", "busy", "authorizing":
                state = .authorizing
                error = nil
            default:
                state = .declined
                error = PathError(code: .decline, message: raw["error"] as? String ?? "Declined", recoverable: false)
            }
        } else if topStatus == "approved" || topStatus == "success" {
            state = defaultState
            error = nil
        } else if topStatus == "refunded" {
            state = .refunded
            error = nil
        } else if topStatus == "cancelled" || topStatus == "canceled" {
            state = .cancelled
            error = nil
        } else if topStatus == "timed_out" {
            state = .timedOut
            error = PathError(code: .timeout, message: raw["error"] as? String ?? "Transaction timed out", recoverable: true)
        } else {
            state = .declined
            error = PathError(code: .decline, message: raw["error"] as? String ?? "Declined", recoverable: false)
        }

        let amount = raw["amount"] as? Int ?? amountFallback
        let currency = raw["currency"] as? String ?? currencyFallback
        let tip = raw["tip"] as? Int
        let txnId = raw["txn_id"] as? String
        let cardLastFour = raw["card_last_four"] as? String
        let receiptAvailable = raw["receipt_available"] as? Bool ?? false
        let now = ISO8601DateFormatter().string(from: Date())
        let resolvedRequestId = (raw["original_req_id"] as? String) ?? (raw["request_id"] as? String) ?? requestId
        return TransactionResult(
            transactionId: txnId,
            requestId: resolvedRequestId,
            state: state,
            amountMinor: amount,
            currency: currency,
            tipMinor: tip,
            cardLastFour: cardLastFour,
            receiptAvailable: receiptAvailable,
            timestampUtc: now,
            error: error
        )
    }

    static func throwIfUnsupportedOperation(_ raw: [String: Any], commandName: String) throws {
        let top = (raw["status"] as? String)?.lowercased() ?? ""
        guard top == "error" else { return }
        let msg = raw["error"] as? String ?? ""
        if msg.contains("Unknown command") {
            throw PathError(
                code: .unsupportedOperation,
                message: "\(commandName) not supported by this terminal firmware",
                recoverable: false
            )
        }
    }
}

File: path-terminal-sdk-android/path-emulator-adapter/src/main/kotlin/tech/path2ai/sdk/emulator/EmulatorWireJsonMapping.kt
package tech.path2ai.sdk.emulator

import kotlinx.serialization.json.*
import tech.path2ai.sdk.core.*
import java.time.Instant

/**
 * Maps emulator wire-protocol JSON responses to [TransactionResult].
 * The wire format uses snake_case keys and string status values.
 */
internal object EmulatorWireJsonMapping {

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Parse a raw JSON response string (after stripping "OK " prefix) into a [TransactionResult].
     */
    fun mapResponse(raw: String, requestId: String): TransactionResult {
        val obj = json.parseToJsonElement(raw).jsonObject

        val status = obj["status"]?.jsonPrimitive?.contentOrNull ?: "error"
        val txnStatus = obj["txn_status"]?.jsonPrimitive?.contentOrNull
            ?: obj["transaction_status"]?.jsonPrimitive?.contentOrNull
            ?: "failed"
        val txnId = obj["txn_id"]?.jsonPrimitive?.contentOrNull
        val amount = obj["amount"]?.jsonPrimitive?.intOrNull ?: 0
        val currency = obj["currency"]?.jsonPrimitive?.contentOrNull ?: "GBP"
        val tip = obj["tip"]?.jsonPrimitive?.intOrNull
        val cardLastFour = obj["card_last_four"]?.jsonPrimitive?.contentOrNull
        val receiptAvailable = obj["receipt_available"]?.jsonPrimitive?.booleanOrNull ?: false
        val errorMessage = obj["error"]?.jsonPrimitive?.contentOrNull

        val state = mapTxnStatus(txnStatus)

        val pathError = if (status == "error" || state == TransactionState.FAILED || state == TransactionState.DECLINED) {
            PathError(
                code = if (state == TransactionState.DECLINED) PathErrorCode.DECLINE else PathErrorCode.TERMINAL_FAULT,
                message = errorMessage ?: "Transaction $txnStatus",
                recoverable = false
            )
        } else null

        return TransactionResult(
            transactionId = txnId,
            requestId = requestId,
            state = state,
            amountMinor = amount,
            currency = currency,
            tipMinor = tip,
            cardLastFour = cardLastFour,
            receiptAvailable = receiptAvailable,
            timestampUtc = Instant.now().toString(),
            error = pathError
        )
    }

    /**
     * Map receipt JSON response to [ReceiptData].
     */
    fun mapReceiptResponse(raw: String, transactionId: String): ReceiptData {
        val obj = json.parseToJsonElement(raw).jsonObject
        val receiptObj = obj["receipt"]?.jsonObject ?: obj

        return ReceiptData(
            transactionId = transactionId,
            requestId = receiptObj["request_id"]?.jsonPrimitive?.contentOrNull,
            merchantReceipt = mapReceiptFields(receiptObj["merchant_receipt"]?.jsonObject ?: JsonObject(emptyMap())),
            customerReceipt = mapReceiptFields(receiptObj["customer_receipt"]?.jsonObject ?: JsonObject(emptyMap())),
            timestampUtc = Instant.now().toString()
        )
    }

    private fun mapReceiptFields(obj: JsonObject): CardReceiptFields {
        return CardReceiptFields(
            copyLabel = obj["copy_label"]?.jsonPrimitive?.contentOrNull ?: "",
            txnType = obj["txn_type"]?.jsonPrimitive?.contentOrNull ?: "",
            amount = obj["amount"]?.jsonPrimitive?.intOrNull ?: 0,
            currency = obj["currency"]?.jsonPrimitive?.contentOrNull ?: "GBP",
            cardScheme = obj["card_scheme"]?.jsonPrimitive?.contentOrNull ?: "",
            maskedPan = obj["masked_pan"]?.jsonPrimitive?.contentOrNull ?: "",
            entryMode = obj["entry_mode"]?.jsonPrimitive?.contentOrNull ?: "",
            aid = obj["aid"]?.jsonPrimitive?.contentOrNull ?: "",
            verification = obj["verification"]?.jsonPrimitive?.contentOrNull ?: "",
            authCode = obj["auth_code"]?.jsonPrimitive?.contentOrNull ?: "",
            merchantId = obj["merchant_id"]?.jsonPrimitive?.contentOrNull ?: "",
            terminalId = obj["terminal_id"]?.jsonPrimitive?.contentOrNull ?: "",
            txnRef = obj["txn_ref"]?.jsonPrimitive?.contentOrNull ?: "",
            timestamp = obj["timestamp"]?.jsonPrimitive?.contentOrNull ?: "",
            status = obj["status"]?.jsonPrimitive?.contentOrNull ?: "",
            retainMessage = obj["retain_message"]?.jsonPrimitive?.contentOrNull
        )
    }

    private fun mapTxnStatus(status: String): TransactionState {
        return when (status.lowercase()) {
            "approved" -> TransactionState.APPROVED
            "declined" -> TransactionState.DECLINED
            "cancelled", "canceled" -> TransactionState.CANCELLED
            "timed_out", "timedout" -> TransactionState.TIMED_OUT
            "failed", "error" -> TransactionState.FAILED
            "processing", "authorizing" -> TransactionState.AUTHORIZING
            "refunded" -> TransactionState.REFUNDED
            "reversed" -> TransactionState.REVERSED
            "pending_device" -> TransactionState.PENDING_DEVICE
            "card_presented" -> TransactionState.CARD_PRESENTED
            "card_read" -> TransactionState.CARD_READ
            else -> TransactionState.FAILED
        }
    }
}
